Lib/_pyrepl/console.py

```diff
@@ -174,7 +174,7 @@ def _excepthook(self, typ, value, tb):
 
     def runsource(self, source, filename="<input>", symbol="single"):
         try:
-            tree = ast.parse(source)
+            tree = self.compile.compiler(source, filename, symbol, ast.PyCF_ONLY_AST)
```

Comment:
Not having the barry flag would fail here if you use `<>` (and even if you use `from __future__ import ...`, since the symbol is). We should find a way to detect the feature without having to make too many passes of compile and AST parsing.

Once you have the AST object, you probably need to walk the top nodes only to detect whether there really was a `__future__` statement or not. But the first thing to do is being able to retrieve an AST node. However, I don't know why `from __future__ import barry_as_FLUFL` does not work:

```
compile('from __future__ import barry_as_FLUFL; 1 <> 2', '', 'exec', flags=ast.PyCF_ONLY_AST)
```

yields

```text
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "", line 1
    from __future__ import barry_as_FLUFL; 1 <> 2
                                             ^^
SyntaxError: invalid syntax
```

I'm not sure if `compile` actually sees `from __future__ import barry_as_FLUFL` when being called like this (it doesn't seem so but I don't know why). 

cc @JelleZijlstra

Reply:
I thought the future import was working. Am I missing something?

Reply:
That's because it automatically inherited already imported flags:

```pycon
>>> from __future__ import barry_as_FLUFL
>>> import ast
>>> compile('1 <> 2', '', 'exec', flags=ast.PyCF_ONLY_AST)
<_ast.Module object at 0x7ff4fde54a90>
>>> compile('1 <> 2', '', 'exec', flags=ast.PyCF_ONLY_AST, dont_inherit=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "", line 1
    1 <> 2
       ^
SyntaxError: invalid syntax
``

Reply:
It behaves the same as on 3.9:

<img width="525" alt="Screenshot 2024-10-06 at 14 07 18" src="https://github.com/user-attachments/assets/46fbb01f-ca84-42b5-9e91-b056d755ce5f">

<img width="837" alt="Screenshot 2024-10-06 at 14 06 52" src="https://github.com/user-attachments/assets/d881d619-9c3a-4474-b2bb-b78b12047e6c">

Reply:
> I thought the future import was working. Am I missing something?

My point was that you probably already had some `from __future__ import barry_as_FLUFL` when you were testing and if you don't set the `dont_inherit` flag, then it would have worked (because it would have inherited the compiler flags).

Now if you put the `from __future__ import barry_as_FLUFL` statement **inside** the `compile()` call, it will not work. `compile` first needs to set the compiler flags before actually compiling the source (I believe). And the reason why it works for scripts in files (not in the REPL) is probably because the parser is already doing something as well. I'm waiting for Jelle's reply on that matter though.

---

Or maybe I misunderstood what you misunderstood?

Reply:
The import is working with the current changes:

<img width="809" alt="Screenshot 2024-10-06 at 14 38 54" src="https://github.com/user-attachments/assets/bc2c8455-ea24-4c38-8790-6c36e7c76283">

So, that's why I'm confused about this discussion. I don't even know what you mean with this "sentence":

> Not having the barry flag would fail here if you use <> (and even if you use from __future__ import ..., since the symbol is).

Also note that compile never supported this in compile:

<img width="689" alt="Screenshot 2024-10-06 at 14 34 57" src="https://github.com/user-attachments/assets/f7cd73d9-e1bf-4193-94ac-1f2195a4b06c">

Reply:
> The import is working with the current changes:

That wasn't my point. My point was:

- How can you detect whether the feature is enabled if you cannot parse the AST?
- What happens if you directly 

> The import is working with the current changes:

The tests are however failing so the fix is incomplete or invalid. In addition, the following does not work:

```pycon
>>> from __future__ import barry_as_FLUFL; 1 <> 2
  File "<python-input-0>", line 1
    from __future__ import barry_as_FLUFL; 1 <> 2
```

The reason why it works is because you are executing `from __future__ import barry_as_FLUFL` first, thereby setting it globally and then when you execute `compile()` the flags are inherited automatically. When you run `from __future__ import barry_as_FLUFL; 1 <> 2` (two statements, one source). But here you don't have any flag set, hence the syntax error. 

The whole point is to make `compile()` work *before* you set the `from __future__ import barry_as_FLUFL`.

Reply:
Note that in 3.12 you could C/C 

```python
from __future__ import barry_as_FLUFL
1 <> 2
```

in the console and it would work because they would be executed as two separate statements. But in the new REPL, this would fail because more than one statements are actually executed:

```pycon
>>> from __future__ import barry_as_FLUFL
... 1 <> 2
...
  File "<python-input-0>", line 1
    from __future__ import barry_as_FLUFL
                                         ^
SyntaxError: multiple statements found while compiling a single statement
```

Reply:
> The tests are however failing so the fix is incomplete or invalid.

This fixes them for me, but I'm not happy with it:

```python
tree = self.compile.compiler(source, filename, "exec", ast.PyCF_ONLY_AST, incomplete_input=False)
```

Reply:
None of the above questions exist in IDLE, maybe we can take a look?
![屏幕截图 2024-10-06 212536](https://github.com/user-attachments/assets/6a50c241-1866-4f1c-8eec-45c92963f40a)
![屏幕截图 2024-10-06 212616](https://github.com/user-attachments/assets/1bc99956-06de-4946-ad28-ee4e1cb9889e)

Reply:
The reason why it works in IDLE is probably because each statement delimited by a `;` (or maybe future-like statements only) or by a newline is executed separately (I don't know if that's the case though, just my gut feeling).

Note that the first example is two different runs(), which is why it works (that's also why it works in 3.12 since we inherit the flags in `compile()` I guess). For the second one... I don't really know (but it probably uses two calls to `compile()` with `1 != 0` being isolated).